core/storage/email/gae_models.py

```diff
@@ -326,9 +326,19 @@ def check_duplicate_message(
             email_hash, sent_datetime_lower_bound=sent_datetime_lower_bound)
 
         for message in messages:
-            if (message.recipient_id == recipient_id and
-                    message.subject == email_subject and
-                    message.html_body == email_body):
+            # We generate hash using recipient_id, email_subject and email_body
+            # in this method. Then we get all the messages who have the same
+            # hash and are sent within the last DUPLICATE_EMAIL_INTERVAL_MINS.
+            # If 'messages' is not None then all its element share the same
```

Comment:
This reasoning isn't quite right (otherwise we'd just check that "messages" exists and be done with it). The gap is that the hash isn't foolproof, it's possible that multiple items have the same hash even though they're different. That's why the additional check here is needed. You might need to simulate this scenario in tests by swapping get_by_hash() with a suitable mock (that simulates a hash collision).

Reply:
Done